Add roon media player integration

pavoni opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change

Roon is a multi-room, music player from [Roon Labs](https://roonlabs.com/)

@marcelveldt did a lot of work creating a python API, and integrating to home assistant - which he has made available as a custom component https://github.com/marcelveldt/roon-hass. 

@marcelveldt had intended to add this to home assistant - but is no longer using roon, and so didn't complete the final step of integrating this component.

With his agreement I intend to do this final integration.

The code in this PR is almost all by @marcelveldt, although I have made minor changes to adjust to HA's flake / lint standards.

I believe the code has been working fo a while (and seems to work well for me - although I'm a new roon user). 

~~There is only one problem I've come across, which is that following integration, and assignment of a roon access token, it is necessary to restart HA. I'm still trying to track down this issue (which is why this PR is WIP).~~ Now fixed

However I'd welcome feedback.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14131

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pavoni committed:
Import roon code.

pavoni committed:
Fix flake8/pylint issues.

pavoni committed:
Fix lint issues, extend timeout, change contact infomation.

pavoni committed:
Add new files to .coveragerc

pavoni committed:
Make file executable.

pavoni committed:
Fix problem with integration not working after initial creation.

pavoni committed:
Improve logic unavailable players by caching data.

MartinHjelmare commented on homeassistant/components/roon/__init__.py:
```diff
@@ -0,0 +1,91 @@
+async def async_setup_entry(hass, entry):
+    """Set up a roonserver from a config entry."""
+    host = entry.data[CONF_HOST]
+    config = hass.data[DATA_CONFIGS].get(host)
+    if config is None:
+        custom_play_action = None
+    else:
+        custom_play_action = config[CONF_CUSTOM_PLAY_ACTION]
+    roonserver = RoonServer(hass, entry, custom_play_action)
```
```suggestion
    roonserver = RoonServer(hass, entry, None)
```

MartinHjelmare commented on homeassistant/components/roon/__init__.py:
```diff
@@ -0,0 +1,91 @@
+    if not await roonserver.async_setup():
+        return False
+
+    hass.data[DOMAIN][host] = roonserver
```
Let's store the server under the `config_entry.entry_id` key instead. That's standard and safer.
```suggestion
    hass.data[DOMAIN][entry.entry_id] = roonserver
```

MartinHjelmare commented on homeassistant/components/roon/__init__.py:
```diff
@@ -0,0 +1,91 @@
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        connections={(CONF_HOST, host)},
```
Ip address is not allowed as a device id as ip address may change.

MartinHjelmare replied:
We should remove connections if we don't have a mac address.
```suggestion
```

MartinHjelmare commented on homeassistant/components/roon/__init__.py:
```diff
@@ -0,0 +1,91 @@
+        identifiers={(DOMAIN, host)},
+        manufacturer="Roonlabs",
+        name=host,
+        model="",
+        sw_version="",
```
Please remove empty items.
```suggestion
```

MartinHjelmare commented on homeassistant/components/roon/__init__.py:
```diff
@@ -0,0 +1,91 @@
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    roonserver = hass.data[DOMAIN].pop(entry.data[CONF_HOST])
```
```suggestion
    roonserver = hass.data[DOMAIN].pop(entry.entry_id)
```

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+"""Config flow for Roon."""
+from asyncio import sleep
+from collections import OrderedDict
```
Dictionaries in Python 3.7+ are ordered by default.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+    }
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class FlowHandler(config_entries.ConfigFlow):
```
This is legacy pattern. See the docs for our current pattern.
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#defining-your-config-flow

```suggestion
class FlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
```

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+class FlowHandler(config_entries.ConfigFlow):
+    """Handle a config flow."""
+
+    user_input = {}
```
We shouldn't set mutable class attributes. That's dangerous. Please use our scaffold script to generate best practice config flow template code.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+        """Handle a flow initialized by the user."""
+        return await self.async_step_init()
+
+    async def async_step_init(self, user_input=None):
```
The init step is legacy as entry point for the config flow. Please use the user step instead.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+
+        return self.async_show_form(step_id="link", errors=errors,)
+
+    async def async_step_import(self, user_input):
+        """Import a config entry.
+
+        Special type of import, we're not actually going to store any data.
+        Instead, we're going to rely on the values that are in config file.
+        """
+        return await self.async_step_user(user_input)
```
This is stale code.
```suggestion
```

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+    async def async_step_init(self, user_input=None):
+        """Confirm the setup."""
+        if user_input is not None:
+            return await self.async_step_link(user_input)
```
Save the host as an instance attribute before moving to the next step instead. Don't pass it as user_input. Each step may have a form part that will be responsible for the user input to that step.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+        if dev_id not in media_players:
+            # new player!
+            media_player = RoonDevice(roon_server, player_data)
+            media_players[dev_id] = media_player
```
Don't store the entity. Just store the `dev_id` in a set.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+            media_player = media_players[dev_id]
+            if media_player and media_player.entity_id:
+                media_player.update_data(player_data)
+                media_player.async_update_callback(media_player.unique_id)
```
Don't access the entity from outside the entity. Signal the entity with the dispatch helper that we already have in place instead.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+            self._media_image_url = None
+
+    async def async_added_to_hass(self):
+        """Register callback."""
```
We should connect the entity state update callback to the signal here.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+    @callback
+    def async_update_callback(self, msg):
+        """Handle device updates."""
+        self.async_schedule_update_ha_state()
```
```suggestion
        self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+        volume = int(volume * 100)
+        try:
+            self._server.roonapi.change_volume(self.output_id, volume)
+        except Exception as exc:  # pylint: disable=broad-except
```
Can we catch a library specific exception instead?

pavoni replied:
I've caught this in a new version of the library. It doesn't need to cascade up to HA.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+        self._server.roonapi.shuffle(self.output_id, shuffle)
+
+    def select_source(self, source):
+        """Select source on player (used to sync/unsync)."""
```
This sounds weird. That select source service should only be used to select a source nothing else. Other tasks need to be done by custom entity services if there's no existing service that is appropriate.

pavoni replied:
Removes for now - can always add something custom later.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,521 @@
+                "media_id": media_id,
+            }
+            _domain, _entity = self._server.custom_play_action.split(".")
+            self.hass.services.call(_domain, _entity, data, blocking=False)
```
We shouldn't do this. We can support a device trigger instead in a follow up PR. See eg the arcam_fmj integration.

pavoni replied:
Again removed for now - can add in a later PR if it's important for people.

MartinHjelmare reviewed (changes requested):
There's a lot of legacy patterns currently specifically around integration set up and config flow. Please use our scaffold script to generate good example code and take it from there.

I've done a first pass review with some comments.

The python-roon repo behind the library is archived. What's the maintenance plan? I don't think we should accept a new integration where the library is already not maintained.

https://github.com/marcelveldt/python-roon

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Update review suggestions

pavoni committed:
Rremove custom play action script.

marcelveldt commented:
@pavoni if you want I can transfer ownership for the lib ?
I wrote this stuff a long time ago and I'm no longer using Roon myself. Feel free to do with it whatever you like.

pavoni commented:
> @pavoni if you want I can transfer ownership for the lib ?
> I wrote this stuff a long time ago and I'm no longer using Roon myself. Feel free to do with it whatever you like.

That would be great - thanks. Happy to take it over.

pavoni committed:
Add test requirements.

pavoni committed:
Tidy manifest.

pavoni committed:
Missed fixes.

pavoni committed:
Refactor config_flow to use current pattern.

pavoni committed:
Add config_flow tests.

pavoni committed:
Refactor to use signal dispatch helpers.

pavoni commented:
> There's a lot of legacy patterns currently specifically around integration set up and config flow. Please use our scaffold script to generate good example code and take it from there.
> 
> I've done a first pass review with some comments.
> 
> The python-roon repo behind the library is archived. What's the maintenance plan? I don't think we should accept a new integration where the library is already not maintained.
> 
> https://github.com/marcelveldt/python-roon

Thanks for taking the time to go through this @MartinHjelmare - the feedback is really appreciated.

I've removed most of the unusual code around play script and syncing zones, and refactored the `config_flow` and `signal dispatch` code.

I still have more work to do:-

- [ ] Remove remaining ghosts from zone sync code
- [ ] Take on maintenance  of the python room library (thanks to @marcelveldt for agreeing to this)
- [ ] Add some roon library exceptions to simplify the media player exception code

Since a lot of the `config_flow` and test logic is new to me - I expect I've made some mistakes in what I've done - but any feedback woule be very welcome as I finish the above.

Thanks again.

pavoni committed:
Remove ToDo: for now.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,102 @@
+    async def authenticate(self) -> bool:
+        """Test if we can authenticate with the host."""
+        try:
+            roonapi = RoonApi(ROON_APPINFO, None, self._host)
```
Make sure to run any blocking I/O in our executor thread pool with `hass.async_add_executor_job`. If the method `authenticate` does more than one call to blocking I/O it's better to make the whole method sync, and schedule it on the executor.

pavoni replied:
This gave me some issues with json serialisation -but in any case the previous `config_flow` code had a non blocking approach (which is probably better because it does a better job of timing out).

So I've reused the non blocking logic of the previous code.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,102 @@
+        if user_input is not None:
+            try:
+                info = await authenticate(self.hass, self._host)
+                return self.async_create_entry(title=DEFAULT_NAME, data=info)
```
Can we get some id from the api to identify the device, besides host? Then we could use that as unique id for the config entry.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

The id needs to be both unique and stable.

pavoni replied:
I've had a dig through the api - and sadly looks like Roon just use the hostname and ip!

MartinHjelmare replied:
Then we need to check the new ip address against the existing entry data ip addresses so that we don't set up the same ip more than once.

MartinHjelmare replied:
Also please move the entry creation outside the try... except block. We can use an `else:` block if needed.

pavoni replied:
@MartinHjelmare Could you give me a pointer on how to get the details of an existing integration. I'm sure it's in the docs - but struggled to find the info.

MartinHjelmare replied:
Call `self._async_current_entries` inside the config flow to get all entries for the integration domain. Then we can check entry.data for whatever item and compare that with some value that we want to filter on.

pavoni replied:
Thanks for the pointer, have done as you suggested.

MartinHjelmare commented on homeassistant/components/roon/manifest.json:
```diff
@@ -0,0 +1,16 @@
+  "requirements": [
+    "roonapi==0.0.20"
+  ],
+  "ssdp": [],
```
Please remove empty items.

MartinHjelmare commented on homeassistant/components/roon/const.py:
```diff
@@ -0,0 +1,38 @@
+    "website": "https://www.home-assistant.io/",
+}
+
+CONFIG_SCHEMA = vol.Schema(
```
The config schema isn't used.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,518 @@
+
+    async def async_added_to_hass(self):
+        """Register callback."""
+        self._remove_signal_status = async_dispatcher_connect(
```
We can use `self.async_on_remove` to register the remove callback.

https://github.com/home-assistant/core/blob/1776540757544981ddb3457a6166681777c3a163/homeassistant/helpers/entity.py#L472-L474

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,518 @@
+            self.async_update_callback,
+        )
+
+    async def async_will_remove_from_hass(self):
```
Then we can remove this.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,518 @@
+        self.async_write_ha_state()
+
+    @property
+    def hidden(self):
```
Hidden is removed from the entity api.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,518 @@
+            "via_hub": (DOMAIN, self._server.host),
+        }
+
+    async def async_update(self):
```
This isn't used since the entity isn't polling.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,518 @@
+        if "source_controls" in self.player_data:
+            for source in self.player_data["source_controls"]:
+                if source["supports_standby"]:
+                    if not source["status"] == "indeterminate":
```
Combine this line with the line above with `and`.

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,156 @@
+        )
+
+        # Initialize Roon background polling
+        ensure_future(self.async_do_loop())
```
We can use `asyncio.create_task` instead.

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,156 @@
+
+    def roonapi_state_callback(self, event, changed_zones):
+        """Callbacks from the roon api websockets."""
+        run_coroutine_threadsafe(
```
If we don't need to wait for the result of the task we can use `hass.add_job` instead.

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,156 @@
+        await self.async_update_changed_players(zone_ids)
+        # check for any removed devices
+        all_devs = {}
+        for zone_id in zone_ids:
```
```py
for zone_id, zone in self.roonapi.zones.items():
```

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,156 @@
+                dev_id = player_data["dev_id"]
+                all_devs[dev_id] = player_data
+        for dev_id in self.all_player_ids:
+            if dev_id not in all_devs.keys():
```
We can remove `.keys()`. Dict membership test uses the keys by default.

MartinHjelmare replied:
Please invert the check and `continue` if true. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,156 @@
+        self.roonapi = None
+        self.all_player_ids = []
+        self.all_playlists = []
+        self.offline_devices = []
```
Please make the containers that should hold unique items sets instead of lists. I suggest changing `all_player_ids` and `offline_devices`.

pavoni committed:
Remove remaining zone / source logic for initial release,

pavoni committed:
Stop authenticate blocking, handle timeout.

pavoni committed:
Removed unneeded code.

pavoni committed:
Review comments update.

pavoni committed:
Fix comment.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+                secs += 5
+                if token:
+                    break
+                await sleep(5, hass.loop)
```
Don't pass the loop. That's deprecated.
```suggestion
                await asyncio.sleep(5)
```

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+"""Config flow for roon integration."""
+from asyncio import sleep
```
Please import asyncio when it's needed. That's standard.
```suggestion
import asyncio
```

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,153 @@
+"""Code to handle the api connection to a Roon server."""
+from asyncio import create_task, sleep
```
```suggestion
import asyncio
```

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,153 @@
+        while not self._exit:
+            await self.async_update_players()
+            # await self.async_update_playlists()
+            await sleep(FULL_SYNC_INTERVAL, self.hass.loop)
```
```suggestion
            await asyncio.sleep(FULL_SYNC_INTERVAL)
```

MartinHjelmare commented on homeassistant/components/roon/server.py:
```diff
@@ -0,0 +1,153 @@
+        )
+
+        # Initialize Roon background polling
+        create_task(self.async_do_loop())
```
```suggestion
        asyncio.create_task(self.async_do_loop())
```

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,461 @@
+"""
+MediaPlayer platform for Roon component.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.roon/
+"""
```
```suggestion
"""MediaPlayer platform for Roon integration."""
```

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,461 @@
+
+from .const import DOMAIN
+
+DEPENDENCIES = ["roon"]
+
```
```suggestion
```

pavoni committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Fix bug in seek.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -374,6 +374,9 @@ def media_previous_track(self):
         self._server.roonapi.seek(self.output_id, position)
+        # Seek doesn't cause an async update - so force one
+        self._media_position = position
+        self.async_write_ha_state()
```
We can't use the async api within sync context. Use `self.schedule_update_ha_state`.
```suggestion
        self.schedule_update_ha_state()
```

pavoni committed:
Use sync rather than async update

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Upgrade library, remove exception now caught in library,

pavoni commented:
@MartinHjelmare I've now taken over the library (and released a version with a small change to catch the seek exception to pypi).

I've also now added some basic documentation. https://github.com/home-assistant/home-assistant.io/pull/14131

So I think it's ready for you to take another look when you have some time.

Really grateful for all the help as I've been catching up on the new HA!

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+                    break
+                await asyncio.sleep(5)
+
+        except Exception as error:  # pylint: disable=broad-except
```
What exception do we expect here? Can we narrow it down?

pavoni replied:
As the roon api stands with`blocking_init=False` I don't think it'll throw.

So I'll remove for now.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+
+        errors = {}
+        if user_input is not None:
+            try:
```
There's no reason to use try... except here. We do that in the next step.

MartinHjelmare commented on homeassistant/components/roon/const.py:
```diff
@@ -0,0 +1,29 @@
+}
+
+
+class RoonException(HomeAssistantError):
```
These exceptions aren't used.

MartinHjelmare commented:
There are some outstanding older comments in the thread also.

pavoni committed:
Review comments.

pavoni committed:
Review changes

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Check for duplicate host before adding.

pavoni commented:
I found an outstanding older comment - now resolved. 

I think I've now changed everything - but obviously let me know if I missed anything.

MartinHjelmare commented on homeassistant/components/roon/strings.json:
```diff
@@ -17,7 +17,8 @@
       "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
-      "unknown": "[%key:common::config_flow::error::unknown%]"
+      "unknown": "[%key:common::config_flow::error::unknown%]",
+      "duplicate_entry": "That entity has already been added."
```
```suggestion
      "duplicate_entry": "That host has already been added."
```

pavoni committed:
Review comment.

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+        return self._last_position_update
+
+    @property
+    def last_changed(self):
```
This isn't used.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+        return self._state
+
+    @property
+    def is_nowplaying(self):
```
This isn't used.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+        return self._shuffle
+
+    @property
+    def repeat(self):
```
This isn't used.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+                    )
+                    break
+        else:
+            return self.media_play()
```
`turn_on` shouldn't return anything.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+
+    def turn_on(self):
+        """Turn on device (if supported)."""
+        if self.supports_standby and "source_controls" in self.player_data:
```
We can invert this check and call media play and then return. Then we can outdent below.

MartinHjelmare commented on homeassistant/components/roon/media_player.py:
```diff
@@ -0,0 +1,452 @@
+
+    def turn_off(self):
+        """Turn off device (if supported)."""
+        if self.supports_standby and "source_controls" in self.player_data:
```
As above.

pavoni committed:
Remove unused code, revise turn_on/turn_off.

MartinHjelmare commented on homeassistant/components/roon/translations/en.json:
```diff
@@ -0,0 +1,27 @@
+        },
+        "error": {
+            "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+            "duplicate_entry": "That entity has already been added.",
```
Make sure to run the translations script in develop mode and commit the changes. This will sync the English default translation with strings.json.
```
python -m script.translations develop
```

pavoni committed:
Sync translations.

MartinHjelmare reviewed (approved):
Nice!

MartinHjelmare commented:
Please add a link to a docs PR in this PR description. Then we can merge.

pavoni commented:
Docs link added.

Thanks for all the time you've spent on this @MartinHjelmare

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+                entry.data[CONF_HOST] for entry in self._async_current_entries()
+            }
+            if self._host in existing:
+                errors["base"] = "duplicate_entry"
```
We need to add a test for this case. Add a mock config entry and start the flow.

https://github.com/home-assistant/core/blob/e32a57ce48c3ac778a5254bf244912081d9d654a/tests/common.py#L658

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
```
Please also add a test for unknown exception.

MartinHjelmare reviewed (changes requested):
Sorry, I noticed we're missing a couple of tests.

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -0,0 +1,82 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.roon.config_flow.RoonHub.authenticate",
```
We should patch `RoonApi` not `RoonHub`.

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+            secs += 5
+            if token:
+                break
+            await asyncio.sleep(5)
```
Probably make 5 a constant so it's easier to patch that in tests.

MartinHjelmare replied:
We'll need to test the timeout case too, that is.

pavoni committed:
Make interim timeout const.

pavoni committed:
Refactor tests.

pavoni committed:
Add tests with an existing config entry.

pavoni commented:
@MartinHjelmare I refactored the existing tests as you suggested, and added a couple of extra tests with an existing config entry.

Sorry I missed them the first time!

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -1,11 +1,28 @@
 from tests.async_mock import patch
 
 
+class roonApiMock:
```
```suggestion
class RoonApiMock:
```

MartinHjelmare commented on homeassistant/components/roon/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+        if user_input is not None:
+            try:
+                info = await authenticate(self.hass, self._host)
+            except CannotConnect:
```
This exception is never raised.

pavoni replied:
Removed.

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -0,0 +1,181 @@
+        "homeassistant.components.roon.const.AUTHENTICATE_TIMEOUT", 0,
+    ), patch(
+        "homeassistant.components.roon.config_flow.RoonApi",
+        return_value=roonApiMock("good_token"),
```
```suggestion
        return_value=RoonApiMock("good_token"),
```

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -0,0 +1,181 @@
+        "homeassistant.components.roon.const.AUTHENTICATE_TIMEOUT", 0,
+    ), patch(
+        "homeassistant.components.roon.config_flow.RoonApi",
+        return_value=roonApiMock(None),
```
```suggestion
        return_value=RoonApiMock(None),
```

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -0,0 +1,181 @@
+        "homeassistant.components.roon.const.AUTHENTICATE_TIMEOUT", 0,
+    ), patch(
+        "homeassistant.components.roon.config_flow.RoonApi",
+        return_value=roonApiMock("good_token"),
```
```suggestion
        return_value=RoonApiMock("good_token"),
```

MartinHjelmare commented on tests/components/roon/test_config_flow.py:
```diff
@@ -0,0 +1,181 @@
+    assert result2["errors"] == {"base": "unknown"}
+
+
+async def test_form_host_already_exists(hass):
```
What is this test meant to be testing? It looks like it's doing a successful entry creation.

pavoni replied:
It's checking the happy path where there is an existing integration. Otherwise the code _might_ always reject a second integration.

MartinHjelmare replied:
Right! I should have read the docstring.

pavoni committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

pavoni committed:
Remove CannotConnect

MartinHjelmare reviewed (approved):
Great!
